docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,446 @@
+In case this is required, the ClusterAddonProvider should take care of this task as well.
+In the case of Helm, we will need to consider that each version of helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.
+
+#### Mapping HelmChartProxy to Helm charts
+
+The Helm artifact repository is huge, and it might happen there are different Helm charts for the same add-on; also in many organizations/products, users are maintaining their own internal Helm repository with certified Helm charts.
+
+In order to provide a solution to this problem, the HelmChartProxy CRD allows to define the link to the Helm chart we want to install. The link to the repository and chart is the canonical, unique source for the Helm chart to install.
+Mapping Kubernetes version to add-ons version
+Cluster add-ons by definition have a lifecycle strictly linked to the Cluster lifecycle, and the main consequences of this is the fact that there is a strict correlation between the Cluster’s Kubernetes version and the add-on version to be installed in a Cluster.
+
+In some cases this correlation is explicit, like e.g. for CPI following the “Versioning Policy for External Cloud Providers”; in other cases the compatibility matrix is defined case by case in documentation, like for “Supported version” in the Calico documentation.
+ 
+On top of that, in many Cluster API installations, the add-on version an operator would like to use is limited to a list of certified versions defined in each organization/product vendor.
+
+We can rely upon existing functionality in Helm (“Capabilities”) to enforce Kubernetes version-specific requirements in a chart, including whether or not to install anything at all.
+
+#### Generating package configurations 
+
+Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.
+
+For ClusterAddonProvider for Helm, a solution is relatively simple. The Helm CLI allows a chart to be configured using a key/value map passed in as CLI arguments, i.e. `--set “key1=value1,key2=value2`. As a result, we can include a similar key/value map in HelmChartProxy for configuring package values.
+
+HelmChartProxy can use an approach similar to templated variables in ClusterClass to fill in variables values. The key/value map can include Go templating representing fields from the Cluster definition, i.e. cluster name, namespace, control plane ref, pod CIDRs. These template fields will be resolved dynamically based on the actual values for each workload cluster selected by HelmChartProxy’s cluster selector.
```

Comment:
Is this how we are proposing that all ClusterAddonProvider's solve this problem? We might not need to prescriptive but I think it's important to avoid re-inventing the wheel. 

With a proposed common solution to a general problem we can do things like provide packages with tools to implement it. This should make implementing new ClusterAddonProvider easier.

Reply:
Not necessarily, I can imagine a situation where other package managers don't configure their values the same way as Helm. The only part I'm prescriptive of is that we'd want to be able to configure the package by referencing fields from the selected Cluster using Go templating.

Reply:
I agree there is value in creating a common solution to a general problem, but at the same time, it seems to me a little bit out of the scope of this proposal, that in some ways, it is exploring this field at this stage.

I think that follow-up iterations of this work should determine what is common/reusable and refactor things accordingly